Implemented the (local) Hilbert symbol for global function fields

S17A05 opened the pull request:
Implemented the computation of the (local) Hilbert symbol at a place of a global function field of odd characteristic following formula 12.4.10 in John Voight's "Quaternion Algebras" (in preparation of extending some quaternion algebra functionality to these global function fields).

The global Hilbert symbol will be implemented via a call to `.is_matrix_ring()` of the appropriate quaternion algebra once the latter method has been extended using an extension of PR #37173, which will be done once both this PR and the referenced PR have been merged.

S17A05 committed:
Implemented the (local) Hilbert symbol for global function fields

Amends: Changed some documentation, adapted characteristic check to check ==2 instead of <= 2

S17A05 committed:
Corrected code in examples

S17A05 committed:
Small optimization and docstring modifications

S17A05 committed:
Docstring corrections

Amend: Added missing multiline continuation markers

S17A05 committed:
Fixed docstrings and small modifications

- Simplified return command, removed `ZZ` from import again
- Return `0` instead of an error for zero inputs

S17A05 committed:
Fixed LINT and docstring issues

S17A05 committed:
Small input adaption, finally fixed docstrings

S17A05 committed:
Merge branch 'develop' into fun_field_hilbert

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,131 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+        Currently only tested for function fields separable over their base
+        since places are not fully supported for other function fields. Only
+        implemented for global function fields of odd characteristic.
+
```
Look here: https://deploy-preview-37554--sagemath.netlify.app/html/en/reference/function_fields/sage/rings/function_field/function_field#sage.rings.function_field.function_field.FunctionField.hilbert_symbol

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,131 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+        - ``a`` and ``b``: Elements of ``self``
+
+        - ``P``: A place of ``self``
+
```
Please see the developer guide to format "INPUT:" section properly.

You may write "elements of this function field" and "a place of this function field".

kwankyu replied:
I meant that `:` should be `--`.  See here: https://doc.sagemath.org/html/en/developer/coding_basics.html#documentation-strings

S17A05 replied:
Sorry, I missed that in my code - it's fixed now.

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,131 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+            ....:     + (2*x/(x^2 + 2))*y + 2*x^2 + 2*x + 1)
+            sage: L.hilbert_symbol(a, b, P)
+            -1
+        """
```
Can you use simpler functions for a, b, c, d? Ideally that can be input in one line.

It would be good to make up an example to demonstrate an arithmetic property of the Hilbert symbol, like the bimultiplicative property.

S17A05 replied:
Thanks for the idea! This actually helped me discover a bug because the application of the residue field map doesn't work as I expected - I fixed it and included random examples for bimultiplicativity.

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,131 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+            raise NotImplementedError('only supported for global function fields')
+
+        if self.characteristic() == 2:
+            raise NotImplementedError('only supported in odd characteristic')
```
You may want to consult this:

https://deploy-livedoc--sagemath.netlify.app/html/en/developer/coding_in_python#exceptions

S17A05 replied:
I am not sure that I understand this comment - is it about the type of error that I am raising here?

kwankyu replied:
Yes. `NotImplementedError` is for the feature not implemented _yet_. Hilbert symbol is defined only for odd characteristic. So there is no feature to be implemented for even characteristic. I am not an expert here. So let me ask if you agree.

S17A05 replied:
You're right, I got something mixed up with the number field case - fixed it.

S17A05 committed:
Implemented reviewer feedback

- Fixed errors caused by unexpected behavior of reduction map
- Added examples for bimultiplicativity
- Corrected docstring

S17A05 committed:
Removed redundant function call in examples

S17A05 committed:
Change characteristic error type

Amend: Fixed weird spacing

S17A05 committed:
Fixed doctest

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,115 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+        # Get the residue field of the completion together with the necssary exponent
+        k = sigma.codomain().base_ring()
+        e = (k.order() - 1) >> 1
+
```
This idiom `x >> 1` for "dividing by 2" is bad. It is obscure. Please replace it with `x // 2`.

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,115 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+        # Finally, put the result together and transform it into the correct output
+        res = k(-1)**(v_a * v_b * e) * a_rd_pw * b_rd_pw
+        return (res.is_one() << 1) - 1
+
```
The above comment applies here too. Moreover this makes the return value a python `int` instead of Sage `Integer`. See https://doc.sagemath.org/html/en/developer/coding_in_python.html#integer-return-values

kwankyu replied:
Here `1 if res.is_one() else  -1` is better. From zen of python: explicit is better than implicit.

S17A05 committed:
Small changes

- Fixed Lint issue
- Implemented reviewer feedback
- Added checks for symmetry

kwankyu commented on src/sage/rings/function_field/function_field.py:
```diff
@@ -1224,6 +1227,121 @@ def completion(self, place, name=None, prec=None, gen_name=None):
+        b0 = ser_b.coefficients()[0]
+
+        # Get the residue field of the completion together with the necssary exponent
+        k = sigma.codomain().base_ring()
```
typo `necessary`

S17A05 committed:
Implemented review feedback

- Rewrote docstring accordings to suggestions
- Fixed typo in comment

S17A05 committed:
Merge branch 'develop' into fun_field_hilbert

S17A05 committed:
Small corrections

kwankyu reviewed (approved):
Thanks. LGTM.
